sktime/utils/_testing/deep_equals.py

```diff
@@ -87,30 +130,47 @@ def _tuple_equals(x, y):
     ----------
     x: tuple or list
     y: tuple or list
+    return_msg : bool, optional, default=False
+        whether to return informative message about what is not equal
 
     Returns
     -------
-    bool - True if x and y are equal in value
+    is_equal: bool - True if x and y are equal in value
         x and y do not need to be equal in reference
+    msg : str, only returned if return_msg = True
+        indication of what is the reason for not being equal
+            concatenation of the following elements:
+            .len - length is not equal
+            [i] - i-th element not equal
     """
+
+    def ret(is_equal, msg):
+        if return_msg:
+            if is_equal:
+                msg = ""
+            return is_equal, msg
+        else:
+            return is_equal
+
     n = len(x)
 
     if n != len(y):
-        return False
+        return ret(False, f".len, x.len = {n} != y.len = {len(y)}")
 
     # we now know dicts are same length
     for i in range(n):
         xi = x[i]
         yi = y[i]
 
         # recurse through xi/yi
-        if not deep_equals(xi, yi):
-            return False
+        is_equal, msg = deep_equals(xi, yi, return_msg=True)
+        if not is_equal:
+            return ret(False, f"[{i}]" + msg)
 
-    return True
+    return ret(True, "")
 
 
-def _dict_equals(x, y):
+def _dict_equals(x, y, return_msg=False):
```

Comment:
Have you tried https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertDictEqual ?

Reply:
that´s not nested, I think? Only looks at the first level. I want to look into data frames and into objects in obj columns, too, since nested data frames occur in the nested_univ mtype representationl

Reply:
also applies only to dict

Reply:
to be clearer, it wouldn´t deal with the recursion, since we need to call `deep_equals` from `_dict_equals`